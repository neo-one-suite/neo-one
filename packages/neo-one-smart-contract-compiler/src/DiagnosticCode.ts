export enum DiagnosticCode {
  GenericUnsupportedSyntax = 0,
  UnknownType,
  UnknownSymbol,
  InvalidBuiltinReference,
  InvalidBuiltinIndex,
  InvalidBuiltinInstanceof,
  InvalidBuiltinImplement,
  InvalidBuiltinModify,
  InvalidBuiltinUsage,
  InvalidContractEvent,
  InvalidContractProperties,
  InvalidContractMethod,
  UnknownCall,
  InvalidSmartContractType,
  InvalidNotifierType,
  InvalidLiteral,
  Invariant,
  MultipleSignatures,
  UnknownModule,
  ReferenceError,
}
