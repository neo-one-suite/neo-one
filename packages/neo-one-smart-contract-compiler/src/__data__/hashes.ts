export const hashes = {
  OLD_NEO_ASSET_HASH: '0xc56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b',
  OLD_GAS_ASSET_HASH: '0x602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7',
};
