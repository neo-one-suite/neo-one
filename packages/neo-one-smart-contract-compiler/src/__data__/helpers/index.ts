export * from './execute';
export * from './extractors';
export * from './startNode';
