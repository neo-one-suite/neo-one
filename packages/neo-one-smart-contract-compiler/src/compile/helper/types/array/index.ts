export * from './ArrayLengthHelper';
export * from './CreateArrayHelper';
export * from './GetArrayIndexHelper';
export * from './IsArrayHelper';
export * from './SetArrayIndexHelper';
export * from './UnwrapArrayHelper';
export * from './WrapArrayHelper';

export * from './typeTests';
