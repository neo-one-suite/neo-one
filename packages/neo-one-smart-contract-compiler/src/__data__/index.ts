import * as helpers from './helpers';

export { helpers };
export * from './keys';
