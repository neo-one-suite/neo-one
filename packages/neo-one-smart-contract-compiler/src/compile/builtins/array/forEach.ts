import { tsUtils } from '@neo-one/ts-utils';
import ts from 'typescript';
import { ScriptBuilder } from '../../sb';
import { VisitOptions } from '../../types';
import { BuiltinInstanceMemberCall } from '../BuiltinInstanceMemberCall';
import { MemberLikeExpression } from '../types';

// tslint:disable-next-line export-name
export class ArrayForEach extends BuiltinInstanceMemberCall {
  public canCall(_sb: ScriptBuilder, _func: MemberLikeExpression, node: ts.CallExpression): boolean {
    return ts.isCallExpression(node) && tsUtils.argumented.getArguments(node).length === 1;
  }

  public emitCall(
    sb: ScriptBuilder,
    func: MemberLikeExpression,
    node: ts.CallExpression,
    optionsIn: VisitOptions,
    visited: boolean,
  ): void {
    const options = sb.pushValueOptions(optionsIn);
    if (!visited) {
      // [arrayVal]
      sb.visit(tsUtils.expression.getExpression(func), options);
    }

    if (tsUtils.argumented.getArguments(node).length < 1) {
      /* istanbul ignore next */
      return;
    }

    // [arr]
    sb.emitHelper(tsUtils.expression.getExpression(func), options, sb.helpers.unwrapArray);
    // [objectVal, arr]
    sb.visit(tsUtils.argumented.getArguments(node)[0], options);
    // []
    sb.emitHelper(node, optionsIn, sb.helpers.arrForEachFunc);
  }
}
