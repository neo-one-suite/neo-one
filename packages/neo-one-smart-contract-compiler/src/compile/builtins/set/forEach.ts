import { tsUtils } from '@neo-one/ts-utils';
import ts from 'typescript';
import { ScriptBuilder } from '../../sb';
import { VisitOptions } from '../../types';
import { BuiltinInstanceMemberCall } from '../BuiltinInstanceMemberCall';
import { CallMemberLikeExpression } from '../types';

// tslint:disable-next-line export-name
export class SetForEach extends BuiltinInstanceMemberCall {
  public canCall(_sb: ScriptBuilder, _func: CallMemberLikeExpression, node: ts.CallExpression): boolean {
    return tsUtils.argumented.getArguments(node).length === 1;
  }

  public emitCall(
    sb: ScriptBuilder,
    func: CallMemberLikeExpression,
    node: ts.CallExpression,
    optionsIn: VisitOptions,
    visited: boolean,
  ): void {
    const options = sb.pushValueOptions(optionsIn);
    if (!visited) {
      // [arrayVal]
      sb.visit(tsUtils.expression.getExpression(func), options);
    }

    if (tsUtils.argumented.getArguments(node).length < 1) {
      /* istanbul ignore next */
      return;
    }

    // [map]
    sb.emitHelper(node, options, sb.helpers.unwrapMap);
    // [keysArr]
    sb.emitOp(node, 'KEYS');
    // [iterator]
    sb.emitSysCall(node, 'System.Iterator.Create');
    // [objectVal, iterator]
    sb.visit(tsUtils.argumented.getArguments(node)[0], options);
    // []
    sb.emitHelper(node, sb.noPushValueOptions(options), sb.helpers.rawEnumeratorForEachFunc({ deserializeKey: true }));

    if (optionsIn.pushValue) {
      // [val]
      sb.emitHelper(node, options, sb.helpers.wrapUndefined);
    }
  }
}
