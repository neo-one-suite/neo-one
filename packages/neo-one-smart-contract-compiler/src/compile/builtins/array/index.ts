import ts from 'typescript';
import { ScriptBuilder } from '../../sb';
import { VisitOptions } from '../../types';
import { BuiltinInstanceOf } from '../BuiltinInstanceOf';
import { BuiltinInterface } from '../BuiltinInterface';
import { Builtins } from '../Builtins';
import { ArrayFilter } from './filter';
import { ArrayForEach } from './forEach';
import { ArrayIterator } from './iterator';
import { ArrayLength } from './length';
import { ArrayMap } from './map';
import { ArrayReduce } from './reduce';

class ArrayInterface extends BuiltinInterface {}
class ArrayValue extends BuiltinInstanceOf {
  public emitInstanceOf(sb: ScriptBuilder, node: ts.Expression, optionsIn: VisitOptions): void {
    const options = sb.pushValueOptions(optionsIn);
    // [val]
    sb.visit(node, options);
    if (optionsIn.pushValue) {
      // [boolean]
      sb.emitHelper(node, options, sb.helpers.isArray);
      // [val]
      sb.emitHelper(node, options, sb.helpers.wrapBoolean);
    } else {
      // []
      /* istanbul ignore next */
      sb.emitOp(node, 'DROP');
    }
  }
}
class ArrayConstructorInterface extends BuiltinInterface {}

// tslint:disable-next-line export-name
export const add = (builtins: Builtins): void => {
  builtins.addInterface('Array', new ArrayInterface());
  builtins.addValue('Array', new ArrayValue());
  builtins.addMember('Array', 'filter', new ArrayFilter());
  builtins.addMember('Array', 'forEach', new ArrayForEach());
  builtins.addMember('Array', '__@iterator', new ArrayIterator());
  builtins.addMember('Array', 'length', new ArrayLength());
  builtins.addMember('Array', 'map', new ArrayMap());
  builtins.addMember('Array', 'reduce', new ArrayReduce());
  builtins.addInterface('ArrayConstructor', new ArrayConstructorInterface());
};
