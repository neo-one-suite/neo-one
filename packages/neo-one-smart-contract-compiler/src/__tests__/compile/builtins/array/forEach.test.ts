import { helpers } from '../../../../__data__';
import { DiagnosticCode } from '../../../../DiagnosticCode';
import ts from 'typescript';
import { ArrayForEach } from '../../../../compile/builtins/array/forEach';

describe('Array.prototype.forEach', () => {
  test('should apply a function over an array', async () => {
    await helpers.executeString(`
      const x = [1, 2, 3];
      let result = 0;
      const y = x.forEach((value) => {
        result += value;
      });

      assertEqual(y, undefined);
      assertEqual(result, 6);
    `);
  });

  test('should apply a function over an array with index', async () => {
    await helpers.executeString(`
      const x = [1, 2, 3];
      let result = 0;
      x['forEach']((value, idx) => {
        result += idx;
      });

      assertEqual(result, 3);
    `);
  });

  test('cannot be referenced', async () => {
    await helpers.compileString(
      `
      const x = [0, 1, 2];
      const y = x.forEach;
    `,
      { type: 'error', code: DiagnosticCode.InvalidBuiltinReference },
    );
  });

  test('cannot be "referenced"', async () => {
    await helpers.compileString(
      `
      const x = [0, 1, 2];
      const y = x['forEach'];
    `,
      { type: 'error', code: DiagnosticCode.InvalidBuiltinReference },
    );
  });

  test('canCall', () => {
    const builtin = new ArrayForEach();
    const expr = ts.createCall(ts.createIdentifier('foo'), undefined, [ts.createIdentifier('bar')]);

    // tslint:disable-next-line no-any
    expect(builtin.canCall(jest.fn() as any, jest.fn() as any, expr)).toBeTruthy();
  });
});
