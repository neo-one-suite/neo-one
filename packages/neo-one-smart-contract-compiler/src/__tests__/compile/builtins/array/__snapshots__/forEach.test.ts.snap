// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array.prototype.forEach cannot be "referenced" 1`] = `
"snippetCode.ts (3,17): Builtin properties cannot be referenced

      1 | 
      2 |       const x = [0, 1, 2];
    > 3 |       const y = x['forEach'];
        |                 ^
      4 |     
"
`;

exports[`Array.prototype.forEach cannot be referenced 1`] = `
"snippetCode.ts (3,17): Builtin properties cannot be referenced

      1 | 
      2 |       const x = [0, 1, 2];
    > 3 |       const y = x.forEach;
        |                 ^
      4 |     
"
`;
